### PR TITLE
test: added tests for Django PG Specific aggregates

### PR DESCRIPTION
**Pull request by gauravsnj**

### CODE REVIEW

File: src/test/java/com/google/cloud/spanner/pgadapter/python/django/DjangoPgAggregates.java [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+          rowBuilder.addValues(value);
+        } else rowBuilder.addValues(Value.newBuilder().setStringValue(rows.get(j)).build());
+      }
+      resultSetBuilder.addRows(rowBuilder.build());
+    }
+    return resultSetBuilder.build();
+  }
+
+  @BeforeClass
+  public static void mockUnnecessarySql() {
```
Comment (olavloite) [inline src/test/java/com/google/cloud/spanner/pgadapter/python/django/DjangoPgAggregates.java]: I think it would be better to add this to the base class instead of here, as this will probably be needed for other Python tests as well (and it cleans up this test, so it only contains code that is relevant for what we are actually testing here)

File: src/test/java/com/google/cloud/spanner/pgadapter/python/django/DjangoPgAggregates.java [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+
+    mockSpanner.putStatementResult(StatementResult.query(Statement.of(unnecessarySql), rs1));
+    mockSpanner.putStatementResult(StatementResult.query(Statement.of(unnecessarySql2), rs1));
+  }
+
+  @Test
+  public void strAggTest() throws IOException, InterruptedException {
+
+    String sql =
+        "SELECT \"singers\".\"firstname\", STRING_AGG(\"singers\".\"firstname\", '|' ) AS \"str\" FROM \"singers\" GROUP BY \"singers\".\"firstname\" LIMIT 21";
```
Comment (olavloite) [inline src/test/java/com/google/cloud/spanner/pgadapter/python/django/DjangoPgAggregates.java]: See this thread for some tips on how to get rid of the `LIMIT 21` clause in these SQL statements: https://stackoverflow.com/questions/60797028/django-annotate-unexpected-limit-21

File: src/test/java/com/google/cloud/spanner/pgadapter/python/django/DjangoPgAggregates.java [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+
+    List<String> result = new ArrayList<>();
```
Comment (olavloite) [inline src/test/java/com/google/cloud/spanner/pgadapter/python/django/DjangoPgAggregates.java]: Same as for the other tests: Let's try to use a more structured data type for this as long as it does not have to pass the border between Java and Python.

File: src/test/java/com/google/cloud/spanner/pgadapter/python/django/DjangoPgAggregates.java [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+    assertEquals(expectedOutput, actualOutput);
+
+    assertEquals(3, mockSpanner.countRequestsOfType(ExecuteSqlRequest.class));
+    assertEquals(sql, mockSpanner.getRequestsOfType(ExecuteSqlRequest.class).get(2).getSql());
+  }
+
+  @Test
+  public void arrAggTest() throws IOException, InterruptedException {
+
+    String sql = "SELECT ARRAY_AGG(\"singers\".\"firstname\" ) AS \"arr\" FROM \"singers\"";
```
Comment (olavloite) [inline src/test/java/com/google/cloud/spanner/pgadapter/python/django/DjangoPgAggregates.java]: Here and elsewhere: This generates a valid PostgreSQL SQL statement, but is this function supported in Spangres?

File: src/test/java/com/google/cloud/spanner/pgadapter/python/django/DjangoPgAggregates.java [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameter;
+import org.junit.runners.Parameterized.Parameters;
+
+@RunWith(Parameterized.class)
+@Category(PythonTest.class)
+public class DjangoPgAggregates extends DjangoTestSetup {
```
Comment (olavloite) [inline src/test/java/com/google/cloud/spanner/pgadapter/python/django/DjangoPgAggregates.java]: nit: Unit test classes should end their name with `Test` (so in this case `DjangoPgAggregatesTest`)

File: src/test/java/com/google/cloud/spanner/pgadapter/python/django/DjangoPgAggregates.java [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+    result1.add("some_column");
+
+    ResultSet rs1 = createResultSet(result1, 1);
+
+    mockSpanner.putStatementResult(StatementResult.query(Statement.of(unnecessarySql), rs1));
+    mockSpanner.putStatementResult(StatementResult.query(Statement.of(unnecessarySql2), rs1));
+  }
+
+  @Test
+  public void strAggTest() throws IOException, InterruptedException {
```
Comment (olavloite) [inline src/test/java/com/google/cloud/spanner/pgadapter/python/django/DjangoPgAggregates.java]: nit (here and elsewhere): the convention is to start test methods with the string `test`. So in this case `testStrAgg`.

File: src/test/python/django/pg_aggregates_test.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    print(Singer.objects.values('firstname').annotate(str=StringAgg('firstname',delimiter = '|')))
+  elif option == 'arr_agg':
+    print(Singer.objects.values('singerid').aggregate(arr=ArrayAgg('firstname')))
+
+
+if __name__ == '__main__':
+
+  if len(sys.argv) < 4:
+    print('Invalid command line arguments')
+    sys.exit()
```
Comment (olavloite) [inline src/test/python/django/pg_aggregates_test.py:55]: nit: `sys.exit(1)`

File: src/test/python/django/test.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+  host = sys.argv[1]
+  port = sys.argv[2]
+
+  try:
+    create_django_setup(host, port)
+    from data.models import Singer
+  except Exception as e:
+    print(e)
+    sys.exit()
```
Comment (olavloite) [inline src/test/python/django/test.py]: nit: `sys.exit(1)`

File: src/test/python/django/test.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
+''' Copyright 2022 Google LLC
```
Comment (olavloite) [inline src/test/python/django/test.py]: Is this file needed? I don't think it's actually being used in this PR.

File: src/test/java/com/google/cloud/spanner/pgadapter/python/django/DjangoPgAggregates.java (hunk 5 above)
Comment (gauravsnj) [reply src/test/java/com/google/cloud/spanner/pgadapter/python/django/DjangoPgAggregates.java]: done

File: src/test/java/com/google/cloud/spanner/pgadapter/python/django/DjangoPgAggregates.java (hunk 6 above)
Comment (gauravsnj) [reply src/test/java/com/google/cloud/spanner/pgadapter/python/django/DjangoPgAggregates.java]: done

File: src/test/java/com/google/cloud/spanner/pgadapter/python/django/DjangoPgAggregates.java (hunk 1 above)
Comment (gauravsnj) [reply src/test/java/com/google/cloud/spanner/pgadapter/python/django/DjangoPgAggregates.java]: done

File: src/test/java/com/google/cloud/spanner/pgadapter/python/django/DjangoPgAggregates.java (hunk 4 above)
Comment (gauravsnj) [reply src/test/java/com/google/cloud/spanner/pgadapter/python/django/DjangoPgAggregates.java]: Yes, I first tested these on a sample database in the cloud spanner via PG Adapter and then only added the unit tests

File: src/test/java/com/google/cloud/spanner/pgadapter/python/django/DjangoPgAggregates.java (hunk 3 above)
Comment (gauravsnj) [reply src/test/java/com/google/cloud/spanner/pgadapter/python/django/DjangoPgAggregates.java]: Yes, I have that action item on my mind.

File: src/test/python/django/pg_aggregates_test.py (hunk 7 above)
Comment (gauravsnj) [reply src/test/python/django/pg_aggregates_test.py:55]: done

File: src/test/python/django/test.py (hunk 9 above)
Comment (gauravsnj) [reply src/test/python/django/test.py]: Thanks for spotting this. I have removed the file

File: src/test/python/django/test.py (hunk 8 above)
Comment (gauravsnj) [reply src/test/python/django/test.py]: done

File: src/test/java/com/google/cloud/spanner/pgadapter/python/django/DjangoPgAggregates.java (hunk 2 above)
Comment (gauravsnj) [reply src/test/java/com/google/cloud/spanner/pgadapter/python/django/DjangoPgAggregates.java]: Thats really helpful. Will keep that in mind for further tests

File: src/test/java/com/google/cloud/spanner/pgadapter/python/django/DjangoTestSetup.java [hunk 10]
```diff
@@ -16,15 +16,139 @@
...
+          rowBuilder.addValues(value);
+        } else rowBuilder.addValues(Value.newBuilder().setStringValue(rows.get(j)).build());
+      }
+      resultSetBuilder.addRows(rowBuilder.build());
+    }
+    return resultSetBuilder.build();
+  }
+
+  @BeforeClass
+  public static void mockUnnecessarySql() {
```
Comment (olavloite) [inline src/test/java/com/google/cloud/spanner/pgadapter/python/django/DjangoTestSetup.java]: nit: Could we rename this to something like `mockDriverInternalSql`? These are actual queries that Django will execute, and without the additional common table expressions for `pg_namespace` and `pg_type`, Django would not work.

**Review by olavloite — APPROVED** [review]
Thanks for applying the changes. LGTM.

File: src/test/java/com/google/cloud/spanner/pgadapter/python/django/DjangoTestSetup.java (hunk 10 above)
Comment (gauravsnj) [reply src/test/java/com/google/cloud/spanner/pgadapter/python/django/DjangoTestSetup.java]: Done. And since these tests were passing on actual database also, it looks like PG Adapter is handling these statements well enough as of now.
